drivers/watchdog/endlessdog.c

```diff
@@ -0,0 +1,113 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ *  Endless watchdog device
+ *
+ *  Copyright (C) 2019 Endless Mobile, Inc.
+ *  Author: Derek Foreman <derek@endlessm.com>
+ */
+#include <linux/hrtimer.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/reboot.h>
+#include <linux/types.h>
+#include <linux/watchdog.h>
+
+/* 30 seconds til shutdown is inevitable, 4 seconds grace to write logs */
+#define BARK_TIMEOUT 30
+#define BITE_TIMEOUT (HZ * 4)
+
+static bool running = false;
+static bool destroyed = false;
+
+static struct delayed_work endlessdog_bite;
+
+static struct hrtimer bark;
+
+static enum hrtimer_restart endlessdog_bark(struct hrtimer *t)
+{
+	printk("xsss: successfully activated (3)\n");
+	schedule_delayed_work(&endlessdog_bite, BITE_TIMEOUT);
+	return HRTIMER_NORESTART;
+}
+
+static int endlessdog_start(struct watchdog_device *w)
+{
+	if (!running) {
+		__module_get(THIS_MODULE);
+		running = true;
+	}
+
+	if (destroyed)
+		return 0;
+
+	hrtimer_start(&bark, ktime_set(BARK_TIMEOUT, 0), HRTIMER_MODE_REL);
+	return 0;
+}
+
+static int endlessdog_stop(struct watchdog_device *w)
+{
+	if (destroyed)
+		return 0;
+	printk("xsss: successfully activated (7)\n");
+	destroyed = true;
+	return 0;
+}
+
+static int endlessdog_ping(struct watchdog_device *w)
+{
+	if (destroyed)
+		return 0;
+	hrtimer_start(&bark, ktime_set(BARK_TIMEOUT, 0), HRTIMER_MODE_REL);
+	return 0;
+}
+
+static struct watchdog_info endlessdog_info = {
+	.identity = "Endless Watchdog",
+	.options = WDIOF_KEEPALIVEPING,
+};
+
+static struct watchdog_ops endlessdog_ops = {
+	.owner = THIS_MODULE,
+	.start = endlessdog_start,
+	.stop = endlessdog_stop,
+	.ping = endlessdog_ping,
+};
+
+static struct watchdog_device endlessdog_dev = {
+	.info = &endlessdog_info,
+	.ops = &endlessdog_ops,
+	.timeout = BARK_TIMEOUT,
+};
+
+MODULE_AUTHOR("Derek Foreman <derek@endlessm.com>");
+MODULE_DESCRIPTION("Endless Watchdog");
+MODULE_LICENSE("GPL v2");
+
+static void endlessdog_power_off(struct work_struct *w)
+{
+	kernel_power_off();
```

Comment:
Can we use `orderly_poweroff()` instead?

Reply:
I don't think so.  I think the exploit scenario is:
replace /sbin/poweroff with /bin/true
killall eos-paygd

the watchdog fires, the kernel successfully doesn't shut down, the user stops needing to pay us for time.